### PR TITLE
Upgrade to JUnit 5

### LINKED ISSUE #9498: Upgrade to JUnit 5
##### **Overview of the feature request**

We should upgrade to JUnit 5 because it's the latest and greatest version of JUnit and JHipster is hip!

##### **Motivation for or Use Case**

Show Java developers the latest and greatest technologies.

##### How

For Maven, exclude `junit:junit` and include `junit-jupiter-engine`:

```xml
<dependency>
    <groupId>org.junit.jupiter</groupId>
    <artifactId>junit-jupiter-engine</artifactId>
    <scope>test</scope>
</dependency>
<dependency>
    <groupId>org.springframework.boot</groupId>
    <artifactId>spring-boot-starter-test</artifactId>
    <scope>test</scope>
    <exclusions>
        <exclusion>
            <groupId>junit</groupId>
            <artifactId>junit</artifactId>
        </exclusion>
    </exclusions>
</dependency>
```

For Gradle:

```groovy
testImplementation ("org.springframework.boot:spring-boot-starter-test") {
    exclude module: "junit"
    exclude group: "com.vaadin.external.google", module: "android-json"
}
testImplementation "org.junit.jupiter:junit-jupiter-engine"
```

In .java files:

1. Replace `org.junit.Before` with `org.junit.jupiter.api.BeforeEach` (or `BeforeAll`, I'm not sure which)
2. Replace `org.junit.Test` with `org.junit.jupiter.api.Test`
3. Remove classes and imports no longer needed:
    1. `import org.junit.runner.RunWith;`
    2. `import org.springframework.test.context.junit4.SpringRunner;`
    3. `@RunWith(SpringRunner.class)`

I'm sure there's more, but this seems like the basics to get started.

### PR DESCRIPTION
**Pull request by hdurix**
Fix #9498

-   Please make sure the below checklist is followed for Pull Requests.

-   [ ] [Travis tests](https://travis-ci.org/jhipster/generator-jhipster/pull_requests) are green
-   [x] Tests are added where necessary
-   [x] Documentation is added/updated where necessary
-   [x] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/master/CONTRIBUTING.md) are followed

### CODE REVIEW

File: generators/server/templates/pom.xml.ejs [hunk 1]
```diff
@@ -363,6 +363,18 @@
             <artifactId>liquibase-mssql</artifactId>
         </dependency>
         <%_ } _%>
+        <dependency>
+            <groupId>org.junit.jupiter</groupId>
+            <artifactId>junit-jupiter-engine</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <%_ if (cucumberTests) { _%>
```
Comment (hdurix) [inline generators/server/templates/pom.xml.ejs]: Cucumber does not support JUnit 5 yet (see https://github.com/cucumber/cucumber-jvm/issues/1149).

JUnit 5 provides a way to use old ("vintage") APIs with junit-vintage, we will be able to remove this after JUnit 5 is added in Cucumber

File: generators/server/templates/pom.xml.ejs [hunk 2]
```diff
@@ -363,6 +363,18 @@
             <artifactId>liquibase-mssql</artifactId>
         </dependency>
         <%_ } _%>
+        <dependency>
```
Comment (hdurix) [inline generators/server/templates/pom.xml.ejs]: @DanielFran should I move the JUnit 5 dependency to https://github.com/jhipster/jhipster?

**Comment by DanielFran** [comment]
@hdurix yes, please.

**Comment by pvliss** [comment]
Awesome work @hdurix. Closing/reopening to relaunch Azure tests after merge of #9539

**Comment by jdubois** [comment]
Great work @hdurix it's incredible we have this for JHipster 6!!

**Comment by hdurix** [comment]
Yes but I don't really understand all the BOM stuff :)

Basically, I've declared in the generated `pom.xml`:
1. add `junit-jupiter-engine`: JUnit 5 dependency, version is defined by Spring Boot
2. add `junit-jupiter-vintage`: JUnit 5 old APIs, only for cucumber
3. remove `junit`: JUnit 4 dependency (added by Spring Boot)

What should be in JHipster dependency and what should be kept in the generated `pom.xml`:
1. `junit-jupiter-engine` added in JHipster BOM (with version or no need?), and also added in the generated `pom.xml`?
2. as it's only for Cucumber, should this only be in the generated `pom.xml`? Or also in JHipster BOM?
3. as I understand, it's easy to set this in JHipster BOM, no need to be specify in the generated `pom.xml`

Sorry @DanielFran I don't really know well how does this work.

**Comment by DanielFran** [comment]
@hdurix In fact no changes is required in Bom, so everything seems ok.

**Comment by DanielFran** [comment]
Thanks @hdurix

**Comment by jdubois** [comment]
@hdurix congrats and don't forget there is a bug bounty!

**Pull request merged**